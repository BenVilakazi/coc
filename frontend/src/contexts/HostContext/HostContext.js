import React, { createContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import useReducerMiddleware from '../useReducerMiddleware';
import HostReducer from './HostReducer';
import socketInstance from '../../socket/socket';
import hostReducerMiddleware from './hostReducerMiddleware';
import config from '../../config';
import AlertModal from '../../components/Modal/AlertModal';

const propTypes = {
  children: PropTypes.node.isRequired,
};

const { emitter } = socketInstance;

const initialState = {
  gameState: 'waiting-for-lobby',
  lobbyID: '',
  players: {},
  playerIDs: [],
  gameSettings: {
    maxPlayers: config.maxPlayers.default,
    winningScore: config.winningScore.default,
    selectedPacks: config.initialSelectedPack,
    handSize: config.handSize,
    winnerScreenDisplayTime: config.winnerScreenDisplayTime,
    hideJoinCode: false,
  },
  deck: { black: [], white: [] },
  loading: [],
  czarSelection: '',
};

export const HostContext = createContext();

function HostProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    hostReducerMiddleware,
    HostReducer,
    initialState,
  );

  const [socketHasError, setSocketHasError] = useState(false);

  function createMessageHandler(currentState) {
    const czarId =
      currentState.playerIDs.length > 0 &&
      currentState.playerIDs.find(
        (playerId) => currentState.players[playerId].isCzar,
      );

    return ({ event, payload, sender }) => {
      switch (event) {
        case 'player-connected': {
          const allPlayers = [...Object.values(currentState.players)];
          const existingPlayer = allPlayers.find(
            (player) =>
              player.name.toUpperCase() === payload.playerName.toUpperCase(),
          );

          if (!existingPlayer) {
            return dispatch({ type: 'PLAYER_CONNECTED', payload });
          }

          if (existingPlayer.status === 'disconnected') {
            return dispatch({ type: 'PLAYER_RECONNECTED', payload });
          }

          return dispatch({ type: 'SEND_NAME_TAKEN_MESSAGE', payload });
        }
        case 'player-disconnected':
          return dispatch({
            type: 'PLAYER_DISCONNECTED',
            payload: {
              ...payload,
              czarId,
            },
          });

        case 'preview-winner':
          return dispatch({ type: 'PREVIEW_WINNER', payload });

        case 'winner-selected':
          return dispatch({ type: 'WINNER_SELECTED', payload });

        case 'lobby-created':
          return dispatch({ type: 'SET_LOBBY_ID', payload });

        case 'black-card-selected':
          return dispatch({ type: 'SELECT_BLACK_CARD', payload });

        case 'join-code-shuffled':
          return dispatch({ type: 'UPDATE_JOIN_CODE', payload });

        case 'player-submit':
          return dispatch({
            type: 'PLAYER_SUBMIT',
            payload: { ...payload, playerId: sender },
          });

        case 'socket-connection-error':
          dispatch({ type: 'UPDATE_JOIN_CODE', payload: { lobbyID: 'ERROR' } });
          return setSocketHasError(true);

        default:
          return undefined;
      }
    };
  }

  useEffect(() => {
    const handleMessage = createMessageHandler(state);
    emitter.on('message', handleMessage);
    return () => {
      emitter.off('message', handleMessage);
    };
  }, [state]);

  useEffect(() => {
    const stagedPlayerIDs = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 'staging',
    );

    const reconnectingPlayersIds = state.playerIDs.filter(
      (playerID) => state.players[playerID].status === 're-connected',
    );

    if (reconnectingPlayersIds.length > 0) {
      const reConnectingPlayers = reconnectingPlayersIds.reduce(
        (acc, id) => ({ ...acc, [id]: state.players[id] }),
        {},
      );

      dispatch({
        type: 'SEND_CARDS_TO_PLAYERS',
        payload: {
          selectedBlackCard: state.selectedBlackCard,
          players: reConnectingPlayers,
          playerIDs: reconnectingPlayersIds,
        },
      });
    }

    if (stagedPlayerIDs.length) {
      const numberOfPlayersExceedingLimit =
        state.playerIDs.length - state.gameSettings.maxPlayers;

      if (numberOfPlayersExceedingLimit > 0) {
        const playerIDsExceedingLimit = state.playerIDs.slice(
          -numberOfPlayersExceedingLimit,
        );

        dispatch({
          type: 'TOO_MANY_PLAYERS',
          payload: {
            players: playerIDsExceedingLimit,
          },
        });
      } else {
        const message =
          state.gameState === 'waiting-for-players'
            ? {
                big: "You've joined the lobby",
                small: 'Please wait for the host to start the game',
              }
            : {
                big: 'A round is in progress',
                small: 'You will join the next round automatically',
              };

        dispatch({
          type: 'SEND_PLAYER_CONNECTED_MESSAGES',
          payload: {
            players: stagedPlayerIDs,
            message,
          },
        });
      }
    }
  }, [state.playerIDs]);

  useEffect(() => {
    // if there is one player AND they are not waiting for the game to begin, end the game
    if (
      state.playerIDs.filter(
        (playerID) => state.players[playerID].status === 'playing',
      ).length === 1 &&
      state.gameState !== 'game-over'
    ) {
      dispatch({
        type: 'GAME_OVER',
        payload: { gameWinner: state.playerIDs[0], playerIDs: state.playerIDs },
      });
    }
  }, [state.playerIDs, state.players]);

  return (
    <HostContext.Provider value={{ state, dispatch }}>
      {children}
      {socketHasError && (
        <AlertModal
          bigText="SOCKET ERROR"
          smallText="Please try again later"
          buttonText="Click anywhere to restart"
          onClick={() => window.location.reload()}
        />
      )}
    </HostContext.Provider>
  );
}

HostProvider.propTypes = propTypes;

export default HostProvider;
