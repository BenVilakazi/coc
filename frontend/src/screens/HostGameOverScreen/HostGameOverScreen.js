import React, { useContext } from 'react';
import styled from 'styled-components';

import { HostContext } from '../../contexts/HostContext/HostContext';
import HostLayout from '../../layouts/HostLayout';
import PlayerList from '../../components/PlayerList/PlayerList';
import InGameSettingsModal from '../../components/HostSettingsMenu/InGameSettingsModal.js';
import Button from '../../components/Buttons/Button';

import WhiteCardsTrio from '../../assets/card-trio-diagonal.svg';
import BlackCardDiagonal from '../../assets/black-card-diagonal.svg';

const LeftPanelWrapper = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;

  .player-list-wrapper {
    overflow-x: auto;
    flex-grow: 1;
  }

  .buttons-wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 32px auto 32px;

    .button.primary {
      font-size: 20px;
      margin-bottom: 16px;
    }
  }
`;

const RightPanelWrapper = styled.div`
  flex: 1;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;

  .winner-display {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin: 40px 0;

    flex: 2;
    text-transform: uppercase;

    .prelude {
      font-size: 28px;
      text-indent: -40px;
      line-height: 1em;
      margin-bottom: -12px;
    }

    .title {
      font-size: 64px;
      font-weight: 900;
      text-align: right;
      line-height: 1em;
    }
  }

  .thanks-message {
    font-size: 46px;
    font-weight: 700;
    text-align: center;
    text-transform: uppercase;
    flex: 1;
  }

  .decorations-wrapper {
    position: relative;
    width: 100%;
    flex: 2;

    .decoration-white-cards,
    .decoration-black-card {
      position: absolute;
      bottom: 0;
    }

    .decoration-white-cards {
      left: 0;
      width: 150px;
    }

    .decoration-black-card {
      right: 0;
      width: 110px;
    }
  }
`;

function LeftPanel() {
  const { state, dispatch } = useContext(HostContext);
  const { players, playerIDs } = state;

  const handleNewGameClick = async () => {
    // TODO: Ideally this would clean up player scores, gameWinner, czarSelection and then replicate the effects seen in HostPregameScreen's Start Button
    await dispatch({ type: 'CLOSE_GAME', payload: {} });
    window.location.reload();
  };

  const handleClickClose = async () => {
    await dispatch({ type: 'CLOSE_GAME', payload: {} });
    window.location.reload();
  };

  return (
    <LeftPanelWrapper>
      <div className="player-list-wrapper">
        <PlayerList playerList={{ players, playerIDs }} />
      </div>
      <div className="buttons-wrapper">
        <Button
          type="button"
          isActive
          className="button primary"
          onClick={handleNewGameClick}
        >
          <p>START NEW GAME</p>
        </Button>

        <Button
          type="button"
          isActive
          className="button"
          onClick={handleClickClose}
        >
          <p>CLOSE GAME</p>
        </Button>
      </div>
    </LeftPanelWrapper>
  );
}

function RightPanel() {
  const { state } = useContext(HostContext);
  const { players, gameWinner } = state;

  return (
    <RightPanelWrapper>
      <h1 className="winner-display">
        <span className="prelude">And the winner is...</span>
        <p className="title" data-testid="winner-name">
          {players[gameWinner].name}
        </p>
      </h1>

      <p className="thanks-message">Thank you for playing!</p>

      <div className="decorations-wrapper">
        <img
          className="decoration-white-cards"
          src={WhiteCardsTrio}
          alt="Three white cards"
        />
        <img
          className="decoration-black-card"
          src={BlackCardDiagonal}
          alt="Diagonal black card"
        />
      </div>
    </RightPanelWrapper>
  );
}

function HostGameOverScreen() {
  return (
    <HostLayout
      className="primary-background"
      left={<LeftPanel />}
      right={<RightPanel />}
      modal={<InGameSettingsModal />}
    />
  );
}

export default HostGameOverScreen;
