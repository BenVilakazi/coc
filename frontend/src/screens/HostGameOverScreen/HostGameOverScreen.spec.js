import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import { HostContext } from '../../contexts/HostContext/HostContext';
import HostGameOverScreen from './HostGameOverScreen';

describe('HostGameOverScreen', () => {
  describe('render', () => {
    const dispatch = jest.fn();

    // Typical expected state in at this point of the game
    const state = {
      gameState: 'showing-winning-cards',
      lobbyID: 'ABCD',
      players: {
        ID1: {
          name: 'foo',
          score: 3,
          isCzar: false,
          submittedCards: [0],
          cards: [
            { text: 'aaaa' },
            { text: 'bbbb' },
            { text: 'cccc' },
            { text: 'dddd' },
          ],
        },
        ID2: {
          name: 'bar',
          score: 1,
          isCzar: true,
          submittedCards: [2],
          cards: [
            { text: 'aaaa' },
            { text: 'bbbb' },
            { text: 'cccc' },
            { text: 'dddd' },
          ],
        },
        ID3: {
          name: 'baz',
          score: 7,
          isCzar: false,
          submittedCards: [1],
          cards: [
            { text: 'eeee' },
            { text: 'ffff' },
            { text: 'gggg' },
            { text: 'hhhh' },
          ],
        },
      },
      selectedBlackCard: {
        text: 'Test Black Card',
        pick: 1,
        pack: 0,
      },
      playerIDs: ['ID1', 'ID2', 'ID3'],
      gameSettings: {
        maxPlayers: 8,
        winningScore: 7,
        selectedPacks: [0],
        handSize: 10,
        winnerScreenDisplayTime: 3000,
      },
      czarSelection: 'ID1',
      gameWinner: 'ID3',
    };

    it('renders', () => {
      const tree = renderer
        .create(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostGameOverScreen />
          </HostContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('renders the name of the game winner', () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostGameOverScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByTestId('winner-name')).toHaveTextContent('baz');
    });

    it('renders "Start new game" button', () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostGameOverScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByText('START NEW GAME')).toBeInTheDocument();
    });

    it('renders "Close game" button', () => {
      render(
        <HostContext.Provider value={{ state, dispatch }}>
          <HostGameOverScreen />
        </HostContext.Provider>,
      );

      expect(screen.queryByText('CLOSE GAME')).toBeInTheDocument();
    });
  });

  describe('functionality', () => {
    const dispatch = jest.fn();

    // Typical expected state in at this point of the game
    const state = {
      gameState: 'showing-winning-cards',
      lobbyID: 'ABCD',
      players: {
        ID1: {
          name: 'foo',
          score: 3,
          isCzar: false,
          submittedCards: [0],
          cards: [
            { text: 'aaaa' },
            { text: 'bbbb' },
            { text: 'cccc' },
            { text: 'dddd' },
          ],
        },
        ID2: {
          name: 'bar',
          score: 1,
          isCzar: true,
          submittedCards: [2],
          cards: [
            { text: 'aaaa' },
            { text: 'bbbb' },
            { text: 'cccc' },
            { text: 'dddd' },
          ],
        },
        ID3: {
          name: 'baz',
          score: 7,
          isCzar: false,
          submittedCards: [1],
          cards: [
            { text: 'eeee' },
            { text: 'ffff' },
            { text: 'gggg' },
            { text: 'hhhh' },
          ],
        },
      },
      selectedBlackCard: {
        text: 'Test Black Card',
        pick: 1,
        pack: 0,
      },
      playerIDs: ['ID1', 'ID2', 'ID3'],
      gameSettings: {
        maxPlayers: 8,
        winningScore: 7,
        selectedPacks: [0],
        handSize: 10,
        winnerScreenDisplayTime: 3000,
      },
      czarSelection: 'ID1',
      gameWinner: 'ID3',
    };

    describe('close game button', () => {
      it('reloads when the close game button is clicked', async () => {
        const { reload } = window.location;

        // window.location properties are read-only, we have to redefine this object to spy on reload
        Object.defineProperty(window, 'location', {
          writable: true,
          value: { ...window.location, reload: jest.fn() },
        });

        render(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostGameOverScreen />
          </HostContext.Provider>,
        );

        await act(async () => userEvent.click(screen.getByText('CLOSE GAME')));

        expect(window.location.reload).toHaveBeenCalledTimes(1);

        window.location.reload = reload;
      });

      it('dispatches CLOSE_GAME event when clicked', async () => {
        const { reload } = window.location;

        // window.location properties are read-only, we have to redefine this object to spy on reload
        Object.defineProperty(window, 'location', {
          writable: true,
          value: { ...window.location, reload: jest.fn() },
        });

        render(
          <HostContext.Provider value={{ state, dispatch }}>
            <HostGameOverScreen />
          </HostContext.Provider>,
        );

        await act(async () => userEvent.click(screen.getByText('CLOSE GAME')));

        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith({
          type: 'CLOSE_GAME',
          payload: {},
        });

        window.location.reload = reload;
      });
    });

    // TODO: Write tests for start new game button once it's properly implemented
  });
});
