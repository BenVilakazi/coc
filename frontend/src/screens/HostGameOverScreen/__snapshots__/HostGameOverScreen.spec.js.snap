// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostGameOverScreen render renders 1`] = `
.c10 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
  background: var(--primary-background-color);
  color: var(--primary-text-color);
}

.c10:hover,
.c10:active,
.c10:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c1 {
  background-image: url(inactiveHamburger.svg);
  background-size: cover;
  background-color: transparent;
  border: 0;
  width: 25px;
  height: 25px;
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .hamburger-container {
  position: absolute;
  right: 17px;
  top: 11px;
  z-index: 10100;
}

.c0 .host-layout-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-layout-header h2 {
  position: relative;
  bottom: -1px;
  font-size: 1rem;
  line-height: 0.8rem;
}

.c0 .components {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c0 .components .left {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 200px;
  background-color: var(--primary-background-color);
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c0 .components .right {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
  min-width: 440px;
}

.c2 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

.c7 {
  stroke: var(--primary-text-color);
}

.c9 {
  stroke: var(--secondary-text-color);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 486px;
  padding: 40px;
  background-color: var(--primary-background-color);
  position: relative;
}

.c4 > div:not(:last-child) {
  margin-bottom: 24px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c5 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 24px 0 26px;
  padding: 7px 0;
  border-bottom: 1px solid #ccc;
  border-radius: 0;
  background-color: var(--primary-background-color);
}

.c5 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--primary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 .player-info span {
  color: var(--primary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.c8 .player-info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 12px 0 26px;
  padding: 7px 12px;
  border-bottom: 0;
  border-radius: 4px;
  background-color: var(--secondary-background-color);
}

.c8 .player-info h1 {
  text-transform: uppercase;
  font-size: 24px;
  font-weight: 900;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  line-height: 1.5em;
  color: var(--secondary-text-color);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8 .player-info span {
  color: var(--secondary-text-color);
  font-size: 24px;
  font-weight: 900;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c3 .player-list-wrapper {
  overflow-x: auto;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 .buttons-wrapper {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 32px auto 32px;
}

.c3 .buttons-wrapper .button.primary {
  font-size: 20px;
  margin-bottom: 16px;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c11 .winner-display {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 40px 0;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  text-transform: uppercase;
}

.c11 .winner-display .prelude {
  font-size: 28px;
  text-indent: -40px;
  line-height: 1em;
  margin-bottom: -12px;
}

.c11 .winner-display .title {
  font-size: 64px;
  font-weight: 900;
  text-align: right;
  line-height: 1em;
}

.c11 .thanks-message {
  font-size: 46px;
  font-weight: 700;
  text-align: center;
  text-transform: uppercase;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c11 .decorations-wrapper {
  position: relative;
  width: 100%;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c11 .decorations-wrapper .decoration-white-cards,
.c11 .decorations-wrapper .decoration-black-card {
  position: absolute;
  bottom: 0;
}

.c11 .decorations-wrapper .decoration-white-cards {
  left: 0;
  width: 150px;
}

.c11 .decorations-wrapper .decoration-black-card {
  right: 0;
  width: 110px;
}

@media (min-width:680px) {
  .c0 .host-layout-header {
    height: 130px;
  }

  .c0 .components .left {
    min-width: 420px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 2.5rem;
    margin-bottom: -4px;
    line-height: 2rem;
  }
}

@media (min-width:1600px) {
  .c0 .components .left {
    min-width: 700px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 3.5rem;
    margin-bottom: -35px;
    line-height: 7rem;
  }
}

@media (min-width:3200px) {
  .c0 .host-layout-header {
    height: 200px;
  }

  .c0 .components .left {
    min-width: 1100px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 6rem;
    margin-bottom: -25px;
    line-height: 7rem;
  }
}

<div
  className="c0 primary-background"
>
  <div
    className="hamburger-container"
  >
    <button
      className="c1"
      onClick={[Function]}
    />
  </div>
  <div
    className="components"
  >
    <div
      className="left"
    >
      <header
        className="c2 host-layout-header"
      >
        <h2>
          CARDS OF CAROUSAL
        </h2>
      </header>
      <div
        className="c3"
      >
        <div
          className="player-list-wrapper"
        >
          <div
            className="c4"
            data-testid="playerList-container"
          >
            <div
              className="c5"
              data-testid="row-foo"
            >
              <img
                alt="card icon"
                data-testid="icon-foo"
                src="offset-white-card-stack-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  foo
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        className="c7"
                        color="primary"
                        d="M32.4285 3.19775V29.5701"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M25.4439 3.51099L23.2217 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M16.5548 3.51099L14.3326 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="c8"
              data-testid="row-bar"
            >
              <img
                alt="card icon"
                data-testid="icon-bar"
                src="black-card-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  bar
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        className="c9"
                        color="secondary"
                        d="M32.4285 3.19775V29.5701"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
            <div
              className="c5"
              data-testid="row-baz"
            >
              <img
                alt="card icon"
                data-testid="icon-baz"
                src="offset-white-card-stack-icon.svg"
                style={
                  Object {
                    "visibility": "visible",
                  }
                }
              />
              <div
                className="player-info"
              >
                <h1>
                  baz
                </h1>
                <div
                  className="c6"
                >
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        className="c7"
                        color="primary"
                        d="M32.4285 3.19775V29.5701"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M25.4439 3.51099L23.2217 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M16.5548 3.51099L14.3326 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M7.66589 3.51099L5.44367 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M0.916168 0.599211L40.9161 29.1692"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                    </g>
                  </svg>
                  <svg
                    data-testid="tally-svg"
                    fill="none"
                    height="30"
                    viewBox="0 0 42 30"
                    width="42"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        className="c7"
                        color="primary"
                        d="M32.4285 3.19775V29.5701"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                      <path
                        className="c7"
                        color="primary"
                        d="M25.4439 3.51099L23.2217 29.5966"
                        data-testid="tally"
                        strokeWidth="2"
                      />
                    </g>
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="buttons-wrapper"
        >
          <button
            className="c10 button primary"
            onClick={[Function]}
            type="button"
          >
            <p>
              START NEW GAME
            </p>
          </button>
          <button
            className="c10 button"
            onClick={[Function]}
            type="button"
          >
            <p>
              CLOSE GAME
            </p>
          </button>
        </div>
      </div>
    </div>
    <div
      className="right"
    >
      <div
        className="c11"
      >
        <h1
          className="winner-display"
        >
          <span
            className="prelude"
          >
            And the winner is...
          </span>
          <p
            className="title"
            data-testid="winner-name"
          >
            baz
          </p>
        </h1>
        <p
          className="thanks-message"
        >
          Thank you for playing!
        </p>
        <div
          className="decorations-wrapper"
        >
          <img
            alt="Three white cards"
            className="decoration-white-cards"
            src="card-trio-diagonal.svg"
          />
          <img
            alt="Diagonal black card"
            className="decoration-black-card"
            src="black-card-diagonal.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
